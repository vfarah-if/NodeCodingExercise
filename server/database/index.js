import mongoose from "mongoose";
import { MongoMemoryServer } from "mongodb-memory-server";
import config  from '../config';

const { defaultMongoUri } = config;
const mongod = new MongoMemoryServer();

export const connect = async (isMemoryServer = true) => {
    const uri = isMemoryServer ? await mongod.getUri() : defaultMongoUri;
    console.debug(`Connecting to MongoDB URI '${uri}'`);

    const connectionOptions = {
        useNewUrlParser: true,
        autoReconnect: true,
        reconnectTries: Number.MAX_VALUE,
        reconnectInterval: 1000
    };

    const result = await mongoose.connect(uri , connectionOptions);
    return result;
};

export const disconnect = async () => {
    await mongoose.connection.dropDatabase();
    await mongoose.connection.close();
    await mongod.stop();
};

export const clearDatabase = async () => {
    const collections = mongoose.connection.collections;

    for (const key in collections) {
        const collection = collections[key];
        await collection.deleteMany();
    }
}
