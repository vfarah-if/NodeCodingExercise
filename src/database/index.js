import mongoose from "mongoose";
import { MongoMemoryServer } from "mongodb-memory-server";
import config from "../config";

const { defaultMongoUri } = config;
const mongod = new MongoMemoryServer();

export const connectDatabase = async (isMemoryServer = true) => {
  const uri = isMemoryServer ? await mongod.getUri() : defaultMongoUri;
  console.debug(`Connecting to MongoDB URI '${uri}'`);

  const connectionOptions = {
    useNewUrlParser: true,
    autoReconnect: true,
    reconnectTries: Number.MAX_VALUE,
    reconnectInterval: 1000,
  };

  const result = await mongoose.connect(uri, connectionOptions);
  return result;
};

export const disconnectAndDropDatabase = async () => {
  await mongoose.connection.dropDatabase();
  await disconnectDatabase();
};

export const disconnectDatabase = async () => {
  await mongoose.connection.close();
  await mongod.stop();
};

export const clearDatabase = async () => {
  const collections = mongoose.connection.collections;

  for (const key in collections) {
    const collection = collections[key];
    await collection.deleteMany();
  }
};

export const removeDatabaseProps = (data) => {
  const result = {
    ...Object.keys(data).reduce((newProperty, key) => {
      if (!key.startsWith("_")) {
        return { ...newProperty, [key]: data[key] };
      }
      return newProperty;
    }, {}),
  };
  return result;
};
