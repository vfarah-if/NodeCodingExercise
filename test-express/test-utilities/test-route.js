import app from "../server/app";
import supertest from "supertest";
const request = supertest(app);

/*
** REMARKS: Help remove duplication in tests and this wraps the 
**          get without needing to include all the imports
*/
export async function getTest(url) {
    const actualResponse = await request.get(url);
    return actualResponse;
}