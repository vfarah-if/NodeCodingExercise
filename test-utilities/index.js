export const cloneCourse = (course) => {
    const stats = { ...course.stats };
    const result = { ...course, stats };
    return result;
  }