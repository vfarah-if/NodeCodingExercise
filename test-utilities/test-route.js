import app from "../src/app";
import supertest from "supertest";
const request = supertest(app);

/*
** REMARKS: Help remove duplication in tests and this wraps the 
**          get without needing to include all the imports
*/
export async function getRequestTest(url) {
    const actualResponse = await request.get(url);
    return actualResponse;
}