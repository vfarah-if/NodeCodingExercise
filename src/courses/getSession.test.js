import { cloneCourse } from "../../test-utilities";
import { createOrUpdateCourse, getSession } from "./index";
import { connectDatabase, disconnectAndDropDatabase } from "../database";

describe("getSession", () => {
  const validCourse = {
    courseId: "04473bf9-6ec6-47e9-be92-77b2bba9b606",
    sessionId: "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    userId: "915a54b7-8220-4c2b-ac03-cc6edda055a5",
    stats: {
      sessionCount: 1,
      totalModulesStudied: 10,
      averageScore: 70,
      timeStudied: 4,
    },
  };

  beforeAll(async () => {
    const isInMemory = true;
    await connectDatabase(isInMemory);
    const course = cloneCourse(validCourse);
    await createOrUpdateCourse(course);
  });

  afterAll(async () => {
    await disconnectAndDropDatabase();
  });

  test("should get expected session from the database", async (done) => {
    const { courseId, userId, sessionId } = validCourse;

    const actual = await getSession(courseId, sessionId, userId);

    expect(actual).toMatchSnapshot();
    done();
  });

  test("should get null when data does not exist", async (done) => {
    const actual = await getSession(
      "non-existent-client-id",
      "non-existent-session-id",
      "non-existent-userid"
    );

    expect(actual).toBeNull();
    done();
  });
});
