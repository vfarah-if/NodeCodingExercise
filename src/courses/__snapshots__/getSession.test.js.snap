// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSession should get NotFoundError when data does not exist 1`] = `[Error: User 'non-ex-userid'not found with course 'non-ex-client-id or session 'non-ex-session-id']`;

exports[`getSession should get expected session from the database 1`] = `
Object {
  "courseId": "04473bf9-6ec6-47e9-be92-77b2bba9b606",
  "sessionId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "stats": Object {
    "averageScore": 70,
    "timeStudied": 4,
    "totalModulesStudied": 10,
  },
  "userId": "915a54b7-8220-4c2b-ac03-cc6edda055a5",
}
`;
